File: src/main/java/com/github/mystery2099/voxlib/optimization/ShapeSimplifier.java
package com.github.mystery2099.voxlib.optimization;

import com.github.mystery2099.voxlib.combination.VoxelAssembly;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.shapes.BooleanOp;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import oshi.util.tuples.Pair;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static java.lang.Math.max;
import static java.lang.Math.min;

public class ShapeSimplifier {

    /**
     * Creates a simplified version of a complex shape by using its bounding box.
     * This is useful for outline shapes that don't need to be as detailed as collision shapes.
     *
     * @param shape The complex shape to simplify.
     * @return A simplified VoxelShape based on the original's bounding box.
     */
    public static VoxelShape simplifyToBoundingBox(VoxelShape shape) {
        return Shapes.create(shape.bounds());
    }

    /**
     * Creates a simplified version of a complex shape by reducing the number of boxes.
     * This method merges boxes that are close to each other or overlap.
     *
     * @param shape    The complex shape to simplify.
     * @param maxBoxes The maximum number of boxes in the simplified shape.
     * @return A simplified VoxelShape with fewer boxes.
     */
    public static VoxelShape simplify(VoxelShape shape, int maxBoxes) {
        AtomicInteger boxCount = new AtomicInteger();
        shape.forAllBoxes((q, w, e, r, t, y) -> boxCount.getAndIncrement());
        if (boxCount.get() <= maxBoxes) return shape;
        List<AABB> boxes = new ArrayList<>();
        shape.forAllBoxes((minx, miny, minz, maxx, maxy, maxz) -> boxes.add(new AABB(minx, miny, minz, maxx, maxy, maxz)));
        while (boxes.size() > maxBoxes) mergeClosestBoxes(boxes);
        AtomicReference<VoxelShape> ret = new AtomicReference<>(Shapes.empty());
        boxes.forEach(box -> ret.set(VoxelAssembly.union(ret.get(), Shapes.create(box))));
        return ret.get();
    }

    public static VoxelShape simplify(VoxelShape shape) {
        return simplify(shape, 8);
    }

    /**
     * Finds the two closest boxes in the list and merges them.
     *
     * @param boxes The list of boxes to process.
     */
    private static void mergeClosestBoxes(List<AABB> boxes) {
        if (boxes.size() <= 1) return;
        Pair<Integer, Integer> closest = null;
        double minDist = Double.MAX_VALUE;
        for (int i = 0; i < boxes.size() - 1; i++)
            for (int j = i + 1; j < boxes.size(); j++) {
                double dist = calculateBoxDistance(boxes.get(i), boxes.get(j));
                if (dist < minDist) {
                    minDist = dist;
                    closest = new Pair<>(i, j);
                }
            }
        if (closest != null) {
            int i = closest.getA();
            int j = closest.getB();
            AABB merged = mergeBoxes(boxes.get(i), boxes.get(j));
            boxes.remove(j);
            boxes.remove(i);
            boxes.add(merged);
        }
    }

    /**
     * Merges two boxes into one that encompasses both.
     *
     * @param box1 The first box.
     * @param box2 The second box.
     * @return A new box that contains both input boxes.
     */
    private static AABB mergeBoxes(AABB box1, AABB box2) {
        return new AABB(
            min(box1.minX, box2.minX),
            min(box1.minY, box2.minY),
            min(box1.minZ, box2.minZ),
            max(box1.maxX, box2.maxX),
            max(box1.maxY, box2.maxY),
            max(box1.maxZ, box2.maxZ)
        );
    }

    /**
     * Calculates the distance between two boxes.
     * Overlapping boxes have a distance of 0.
     *
     * @param box1 The first box.
     * @param box2 The second box.
     * @return The distance between the boxes.
     */
    private static double calculateBoxDistance(AABB box1, AABB box2) {
        if (box1.intersects(box2)) return 0;
        double dx = max(0.0, max(box1.minX - box2.maxX, box2.minX - box1.maxX));
        double dy = max(0.0, max(box1.minY - box2.maxY, box2.minY - box1.maxY));
        double dz = max(0.0, max(box1.minZ - box2.maxZ, box2.minZ - box1.maxZ));
        return dx * dx + dy * dy + dz * dz;
    }


    /**
     * Creates a simplified outline shape for a block with the given dimensions.
     * This creates a hollow box shape that's more efficient than a complex shape.
     *
     * @param minX The minimum X coordinate.
     * @param minY The minimum Y coordinate.
     * @param minZ The minimum Z coordinate.
     * @param maxX The maximum X coordinate.
     * @param maxY The maximum Y coordinate.
     * @param maxZ The maximum Z coordinate.
     * @param thickness The thickness of the outline (default is 1).
     * @return A simplified hollow box VoxelShape.
     */
    public static VoxelShape createOutlineShape(
        double minX, double minY, double minZ,
        double maxX, double maxY, double maxZ,
        double thickness
    ) {
        VoxelShape outer = Shapes.box(minX, minY, minZ, maxX, maxY, maxZ);
        if(maxX - minX <= 2 * thickness ||
            maxY - minY <= 2 * thickness ||
            maxZ - minZ <= 2 * thickness
        ) return outer;
        VoxelShape inner = Shapes.box(
            minX + thickness, minY + thickness, minZ + thickness,
            maxX - thickness, maxY - thickness, maxZ - thickness
        );
        return Shapes.join(outer, inner, BooleanOp.ONLY_FIRST);
    }

    public static VoxelShape createOutlineShape(
        double minX, double minY, double minZ,
        double maxX, double maxY, double maxZ
    ) { return createOutlineShape(minX, minY, minZ, maxX, maxY, maxZ, 1); }
}

File: src/main/kotlin/com/github/mystery2099/voxlib/optimization/ShapeSimplifier.kt
package com.github.mystery2099.voxlib.optimization

import com.github.mystery2099.voxlib.combination.VoxelAssembly.createCuboidShape
import net.minecraft.util.math.Box
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes

/**
 * A utility class for simplifying complex VoxelShapes to improve performance.
 *
 * Complex shapes with many boxes can cause performance issues, especially
 * when used for outlines that change frequently. This class provides methods
 * to create simplified versions of shapes for use in outlines.
 */
object ShapeSimplifier {

    /**
     * Creates a simplified version of a complex shape by using its bounding box.
     * This is useful for outline shapes that don't need to be as detailed as collision shapes.
     *
     * @param shape The complex shape to simplify.
     * @return A simplified VoxelShape based on the original's bounding box.
     */
    fun simplifyToBoundingBox(shape: VoxelShape): VoxelShape {
        if (shape.isEmpty) return VoxelShapes.empty()

        val boundingBox = shape.boundingBox
        return VoxelShapes.cuboid(boundingBox)
    }

    /**
     * Creates a simplified version of a complex shape by reducing the number of boxes.
     * This method merges boxes that are close to each other or overlap.
     *
     * @param shape The complex shape to simplify.
     * @param maxBoxes The maximum number of boxes in the simplified shape.
     * @return A simplified VoxelShape with fewer boxes.
     */
    fun simplify(shape: VoxelShape, maxBoxes: Int = 8): VoxelShape {
        require(maxBoxes >= 1) { "maxBoxes must be at least 1" }

        val boxes = mutableListOf<Box>()
        shape.forEachBox { minX, minY, minZ, maxX, maxY, maxZ ->
            boxes.add(Box(minX, minY, minZ, maxX, maxY, maxZ))
        }

        if (boxes.size <= maxBoxes) return shape

        // Mid-size inputs use the Java priority-queue merger; tiny/huge inputs use
        // the O(n²) scan (avoids queue candidate storage when n is very large).
        val simplifiedBoxes = if (boxes.size in PRIORITY_QUEUE_BOX_RANGE) {
            mergeClosestBoxesWithQueue(boxes, maxBoxes)
        } else {
            mergeClosestBoxesWithScan(boxes, maxBoxes)
        }

        return simplifiedBoxes.fold(VoxelShapes.empty()) { acc, box ->
            VoxelShapes.union(acc, VoxelShapes.cuboid(box))
        }
    }

    private fun mergeClosestBoxesWithQueue(boxes: List<Box>, maxBoxes: Int): List<Box> =
        DeterministicBoxMerger.mergeClosest(boxes, maxBoxes)

    private fun mergeClosestBoxesWithScan(boxes: MutableList<Box>, maxBoxes: Int): List<Box> {
        while (boxes.size > maxBoxes) {
            mergeClosestPairWithScan(boxes)
        }
        return boxes
    }

    private fun mergeClosestPairWithScan(boxes: MutableList<Box>) {
        var closestFirst = 0
        var closestSecond = 1
        var minimumDistance = Double.MAX_VALUE
        for (i in 0 until boxes.size - 1) {
            for (j in i + 1 until boxes.size) {
                val distance = calculateBoxDistance(boxes[i], boxes[j])
                if (distance < minimumDistance) {
                    minimumDistance = distance
                    closestFirst = i
                    closestSecond = j
                }
            }
        }

        val mergedBox = mergeBoxes(boxes[closestFirst], boxes[closestSecond])
        boxes.removeAt(closestSecond)
        boxes.removeAt(closestFirst)
        boxes.add(mergedBox)
    }

    /**
     * Calculates the distance between two boxes.
     * Overlapping boxes have a distance of 0.
     *
     * @param box1 The first box.
     * @param box2 The second box.
     * @return The distance between the boxes.
     */
    private fun calculateBoxDistance(box1: Box, box2: Box): Double {
        val dx = maxOf(0.0, maxOf(box1.minX - box2.maxX, box2.minX - box1.maxX))
        val dy = maxOf(0.0, maxOf(box1.minY - box2.maxY, box2.minY - box1.maxY))
        val dz = maxOf(0.0, maxOf(box1.minZ - box2.maxZ, box2.minZ - box1.maxZ))

        return dx * dx + dy * dy + dz * dz
    }

    /**
     * Merges two boxes into one that encompasses both.
     *
     * @param box1 The first box.
     * @param box2 The second box.
     * @return A new box that contains both input boxes.
     */
    private fun mergeBoxes(box1: Box, box2: Box): Box {
        return Box(
            minOf(box1.minX, box2.minX),
            minOf(box1.minY, box2.minY),
            minOf(box1.minZ, box2.minZ),
            maxOf(box1.maxX, box2.maxX),
            maxOf(box1.maxY, box2.maxY),
            maxOf(box1.maxZ, box2.maxZ)
        )
    }

    /**
     * Creates a simplified outline shape for a block with the given dimensions.
     * This creates a hollow box shape that's more efficient than a complex shape.
     *
     * @param minX The minimum X coordinate.
     * @param minY The minimum Y coordinate.
     * @param minZ The minimum Z coordinate.
     * @param maxX The maximum X coordinate.
     * @param maxY The maximum Y coordinate.
     * @param maxZ The maximum Z coordinate.
     * @param thickness The thickness of the outline (default is 1).
     * @return A simplified hollow box VoxelShape.
     */
    fun createOutlineShape(
        minX: Number, minY: Number, minZ: Number,
        maxX: Number, maxY: Number, maxZ: Number,
        thickness: Number = 1
    ): VoxelShape {
        val t = thickness.toDouble()
        val minXd = minX.toDouble()
        val minYd = minY.toDouble()
        val minZd = minZ.toDouble()
        val maxXd = maxX.toDouble()
        val maxYd = maxY.toDouble()
        val maxZd = maxZ.toDouble()

        val outerBox = createCuboidShape(minXd, minYd, minZd, maxXd, maxYd, maxZd)
        if (maxXd - minXd <= 2 * t || maxYd - minYd <= 2 * t || maxZd - minZd <= 2 * t) {
            return outerBox
        }

        val innerBox = createCuboidShape(
            minXd + t, minYd + t, minZd + t,
            maxXd - t, maxYd - t, maxZd - t
        )
        return VoxelShapes.combineAndSimplify(
            outerBox, innerBox,
            net.minecraft.util.function.BooleanBiFunction.ONLY_FIRST
        )
    }

    /** Box counts where the priority-queue merger wins; outside this, use the scan. */
    private val PRIORITY_QUEUE_BOX_RANGE = 96..256
}
